### PR TITLE
Use pure-Ruby Fiber-based Enumerator#next

### PR DESCRIPTION
**Pull request by headius**
This code has been passed around multiple implementations as a
pure-Ruby version of Enumerator#next that just uses fibers. We
have had many issues with the separate fiber-like impl we used
for Enumerator#next, and the logical path forward is to unify that
with Fiber via this code.

There are a few caveats to this migration:

* Three new introduced failures appear to be shared with other
  impls using this logic (rbx, truffleruby).
* This commit breaks the fast-path Array logic for Enumerator#next
  so it's always going through a thread. However the protocol here
  would allow us to implement #to_generator on any object, which
  should make it possible to short-circuit more easily (in Ruby!).
* I have not measured performance. However the logic surrounding
  the Fiber calls should not be particularly heavy.
* I have not vetted the cleanup side of this, except to confirm
  that this implementation seems about as good at jruby/jruby#5671
  in that it does appear to GC, but for a high-churn benchmark it
  eventually collapses without explicit GCs. So...no worse?

This code is modified from TruffleRuby's copy of Rubinius's code,
with Rubinius's license being very liberal and TruffleRuby's being
identical to JRuby's.

I am proposing this for 9.2.7 because it seems like mostly a lateral move, although there were some tag changes. There are at least two todos before release:

* Green build obviously.
* Implement at least a few "fast" generators for core collections like Array and Hash.

### CODE REVIEW

File: core/src/main/java/org/jruby/RubyEnumerator.java [hunk 1]
```diff
@@ -519,91 +524,26 @@ public IRubyObject with_index19(ThreadContext context, IRubyObject arg, final Bl
...
-        return ensureNexter(getRuntime()).hasNext();
+        ThreadContext context = runtime.getCurrentContext();
+        return sites(context).next_p.call(context, this, generator).isTrue();
     }
 
     @Override
     public Object next() {
-        return next( getRuntime().getCurrentContext() ).toJava( java.lang.Object.class );
+        ThreadContext context = getRuntime().getCurrentContext();
+        return sites(context).next_p.call(context, this, this).toJava( java.lang.Object.class );
```
Comment (kares) [inline core/src/main/java/org/jruby/RubyEnumerator.java]: you might have wanted `sites(context).next.call(...)` here to invoke `next` (not `next?`)

File: core/src/main/ruby/jruby/kernel/enumerator.rb [hunk 2]
```diff
@@ -7,6 +7,100 @@ class Yielder
...
+    end
+
+    begin
+      return @__generator__.next if @__generator__.next?
+    rescue StopIteration
+      nil # the enumerator could change between next? and next leading to StopIteration
+    end
+
+    exception = StopIteration.new 'iteration reached an end'
+    JRuby.ref(exception).result = @__generator__.result
```
Comment (kares) [inline core/src/main/ruby/jruby/kernel/enumerator.rb]: a bit ugly maybe worth extending `StopIteration.new` 
or having a `StopIteration.__new__('msg', result)`?
... you wouldn't need to have `RubyStopIteration#setResult` around

Comment (headius) [reply core/src/main/ruby/jruby/kernel/enumerator.rb]: Yeah this is troublesome. The result is not supposed to be mutable through normal Ruby APIs, if we want to be strictly compatible, but there's also no way to construct a StopIteration with a specific result from Ruby. Both the Rubinius and TruffleRuby versions of this code do something basically the same. I went with what was simplest for now.

File: core/src/main/java/org/jruby/RubyEnumerator.java (hunk 1 above)
Comment (headius) [reply core/src/main/java/org/jruby/RubyEnumerator.java]: You are correct sir, good catch. Must not be tested.

Comment (headius) [reply core/src/main/java/org/jruby/RubyEnumerator.java]: Looks like this is tested and failed. I'll finish up the PR and make sure it's all green.

**Pull request merged**